File: source-code/risk-analysis/capital-asset-pricing-model.js
// +------------------------------------------------------------+
//   Model: Capital Asset Pricing Model (CAPM)					
//   © Copyright: https://discountingcashflows.com	
// +------------------------------------------------------------+

Input(
  {
    _MARKET_PREMIUM: '',
    _RISK_FREE_RATE: '',
    BETA: '',
  }
);

$.when(
  get_profile(),
  get_treasury(),
  get_fx(),
  get_risk_premium()).done(
  function(_profile, _treasury, _fx, _risk_premium){
    var response = new Response({
      profile: _profile,
      treasury: _treasury,
      risk_premium: _risk_premium,
    });
    
    // +---------------- ASSUMPTIONS SECTION -----------------+
    setAssumption('_MARKET_PREMIUM', response.risk_premium.totalEquityRiskPremium );
	// Set beta (used in calculating the discount rate)
    if(response.profile.beta){
    	setAssumption('BETA', response.profile.beta);
    }
    else{
    	setAssumption('BETA', 1);
    }
    // Risk free rate is the yield of the 10 year treasury note
	setAssumption('_RISK_FREE_RATE', response.treasury.year10);
	
    // Expected Return
    var expectedReturn = getAssumption('_RISK_FREE_RATE') + getAssumption('BETA')*getAssumption('_MARKET_PREMIUM');
    
    // If we are calculating the value per share for a watch, we can stop right here.
    if(_StopIfWatch(toP(expectedReturn), '%')){
      return;
    }
    
    print(expectedReturn, "Cost of Equity", '%');
});

Description(`
	<h5>Capital Asset Pricing Model (CAPM)</h5> 
	The Capital Asset Pricing Model describes the relationship between risk and expected return for assets, particularly stocks. Its formula is used to calculate the cost of equity, or the rate of return a company is expected to pay to equity investors.
	<p class='text-center'>Read more: <a href='https://github.com/DiscountingCashFlows/Documentation/blob/main/models-documentation/dividend-discount-models.md#discount-rate-cost-of-equity' target='_blank'><i class="fab fa-github"></i> GitHub Documentation</a></p>
`);

File: source-code/risk-analysis/capital-asset-pricing-model.py
"""
    Model: Capital Asset Pricing Model (CAPM)

    © Copyright discountingcashflows.com
"""

# Render model description
model.render_description({
    "data": r"""
    ## Capital Asset Pricing Model (CAPM)
    
    The Capital Asset Pricing Model describes the relationship between risk and expected return for assets, particularly stocks. Its formula is used to calculate the cost of equity, or the rate of return a company is expected to pay to equity investors.
    
    Read more: [GitHub Model Documentation](https://github.com/DiscountingCashFlows/Documentation/blob/main/models-documentation/dividend-discount-models.md#discount-rate-cost-of-equity)
    """,
    "properties": {
        "width": "responsive"
    }
})

# Initialize assumptions
assumptions.init({
    "data": {
        "beta": data.get("profile:beta", default=1),
        "%risk_free_rate": data.get("treasury:year10"),
        "%market_premium": data.get("risk:totalEquityRiskPremium"),
    }
})

# Calculate the expected return
expected_return = assumptions.get("%risk_free_rate") + assumptions.get("beta") * assumptions.get("%market_premium")

model.set_final_value({
    "value": expected_return,
    "units": '%'
})

# Set description for assumptions
assumptions.set_description({
    "beta": r"""
        ## Beta

        Beta is a value that measures the price fluctuations (volatility) of a stock with respect to fluctuations in the overall stock market.
    """,

    "%risk_free_rate": r"""
        ## Risk-Free Rate

        The risk-free rate represents the interest an investor would expect from an absolutely risk-free investment over a specified period of time.
        By default, it is equal to the current yield of the U.S. 10 Year Treasury Bond.
    """,

    "%market_premium": r"""
        ## Market Premium

        Market risk premium represents the excess returns over the risk-free rate that investors expect for taking on the incremental risks connected to the equities market.
    """,
})

print("End of model.")
